File: protobuf_prototype/cpp/Delay.cpp
#include "Delay.hpp"

#include <sstream>

const long Delay::tau = 500000000;

Delay::Delay() : _distribution(0, tau)
{
  _rng.seed(std::random_device()());
}

Delay::operator bool() const
{
  return (0 < tv_sec || 0 < tv_nsec);
}

void Delay::operator()()
{
  ::nanosleep(this, NULL);
}

double Delay::seconds() const
{
  return ((1.e9 * tv_sec) + tv_nsec) / 1.e9;
}

void Delay::reset()
{
  tv_sec = 0;
  tv_nsec = 0;
}

void Delay::randomize()
{
  tv_nsec = tau + _distribution(_rng);
}

void Delay::parse(const std::string& str)
{
  std::istringstream iss(str);
  tv_nsec = 0L;
  iss >> tv_nsec;
  std::string units;
  iss >> units;
  if (units == "s") {
    tv_nsec *= 1000000000;
  }
  else if (units == "u") {
    tv_nsec *= 1000;
  }
}

File: protobuf_prototype/csharp/src/Delay.cs
using System;
using System.Threading;

namespace protobuf_prototype_csharp
{
	public class Delay
	{
		private const int maximum = 500;

		private int milliSeconds = 0;

		private Random random = new Random();

		public int Seconds
		{
			get { return 1000 * milliSeconds; }
			set { milliSeconds = value / 1000; }
		}

		public Delay()
		{
			// NOP
		}

		public void Reset()
		{
			milliSeconds = 0;
		}

		public void Randomize()
		{
			milliSeconds = random.Next(0, maximum);
		}

		public void Parse(String str)
		{
			String units = "m";
			if (0 < str.Length)
			{
				units = str.Substring(str.Length - 1);
				str.Remove(str.Length - 1);
			}

			milliSeconds = Int32.Parse(str);
			if (units == "s")
			{
				milliSeconds *= 1000;
			}
			else if (units == "u")
			{
				milliSeconds /= 1000;
			}
		}

		public void Execute()
		{
			if (maximum < milliSeconds)
			{
				Console.WriteLine("WARNING: Sleeping {0} milli-seconds...", milliSeconds);
			}
			Thread.Sleep(milliSeconds);
		}
	}
}
